### ISSUE TITLE
Native support for Focal Loss

### ISSUE THREAD

**Issue by GLevV** [microsoft/LightGBM#3706]
Could you add native support for Focal Loss? There are implementations for LGBM and XGB. They show better performance on imbalanced data, but they are slow and clunky. Also allow to use MSE for classification (also known as [square loss](https://en.wikipedia.org/wiki/Loss_functions_for_classification#Square_loss), it's a proper scorer function and in some cases could be better than LogLoss). More classification losses could be added to stand out in this department (since all GBDTs algorithms use LogLoss and nothing else).

**Comment by guolinke** [comment]
thank you, a good feature request!

For use regression objective for classification, I think LightGBM already supports it. 
You can use the LightGBM native API, not sklearn API. And then set objective to l2, metric to classification metrics. I think it should work.

**Comment by guolinke** [comment]
@btrotta @shiyu1994 are you interesting in implementing Focal loss?

**Comment by StrikerRUS** [comment]
@guolinke 
> You can use the LightGBM native API, not sklearn API.

I think sklearn API can be used as well, but the most general class `LGBMModel` from it.
```
lgb.LGBMModel(objective='l2', metric='binary_logloss').fit(X, y)
```

**Comment by shiyu1994** [comment]
I can implement it.

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.
